src/net/server/service.rs

```diff
@@ -0,0 +1,624 @@
+impl<RequestOctets, Target, Future, F> Service<RequestOctets> for F
+where
+    F: Fn(
+        Request<Message<RequestOctets>>,
+    ) -> Result<
+        Transaction<
+            Result<CallResult<RequestOctets, Target>, ServiceError>,
+            Future,
+        >,
+        ServiceError,
+    >,
+    RequestOctets: AsRef<[u8]>,
+    Target: Composer + Default + Send + Sync + 'static,
+    Future: std::future::Future<
+            Output = Result<CallResult<RequestOctets, Target>, ServiceError>,
+        > + Send,
+{
+    type Target = Target;
+    type Future = Future;
+
+    fn call(
+        &self,
+        message: Request<Message<RequestOctets>>,
+    ) -> Result<
+        Transaction<
+            Result<CallResult<RequestOctets, Self::Target>, ServiceError>,
+            Self::Future,
+        >,
+        ServiceError,
+    > {
+        (*self)(message)
+    }
+}
+
+//------------ ServiceError --------------------------------------------------
+
+/// An error reported by a [`Service`].
+#[derive(Debug)]
+pub enum ServiceError {
+    /// The service was unable to parse the request.
+    FormatError,
+
+    /// The service encountered a service-specific error condition.
+    InternalError,
+
+    /// The service was unable to assemble the response.
+    NotImplemented,
+
+    /// The service declined to handle the request.
+    Refused,
+}
+
+impl ServiceError {
+    pub fn rcode(&self) -> Rcode {
+        match self {
+            Self::FormatError => Rcode::FormErr,
+            Self::InternalError => Rcode::ServFail,
+            Self::NotImplemented => Rcode::NotImp,
+            Self::Refused => Rcode::Refused,
+        }
+    }
+}
+
+//--- Display
+
+impl Display for ServiceError {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        match self {
+            Self::FormatError => write!(f, "Format error"),
+            Self::InternalError => write!(f, "Internal error"),
+            Self::NotImplemented => write!(f, "Not implemented"),
+            Self::Refused => write!(f, "Refused"),
+        }
+    }
+}
+
+//--- From<PushError>
+
+impl From<PushError> for ServiceError {
+    fn from(_: PushError) -> Self {
+        Self::InternalError
+    }
+}
+
+//--- From<ParseError>
+
+impl From<ParseError> for ServiceError {
+    fn from(_: ParseError) -> Self {
+        Self::FormatError
+    }
+}
+
+//------------ ServerCommand ------------------------------------------------
+
+/// Command a server to do something.
+#[derive(Copy, Clone, Debug)]
+pub enum ServerCommand<T: Sized> {
+    #[doc(hidden)]
+    /// This command is for internal use only.
+    Init,
+
+    /// Command the server to alter its configuration.
+    Reconfigure(T),
+
+    /// Command the connection handler to terminate.
+    ///
+    /// This command is only for connection handlers for connection-oriented
+    /// transport protocols, it should be ignored by servers.
+    CloseConnection,
```

Comment:
`ServerCommand::CloseConnection` is unusd and should be removed. It's a left over from when `ServiceFeedback` and `ServerCommand` in earlier code were one and the same rather than split as they are now.

`ServiceFeedback::CloseConnection` could be used by a `Service` or middleware processor that detects that a client connection is behaving maliciously or beyond some limit in order to forcibly terminate the connection to the bad actor.

Reply:
That seems counter to what is said in [RFC 7766](https://datatracker.ietf.org/doc/html/rfc7766) which gives the following scenarios under which a connection could be closed by the server:

- "If the server needs to close a dormant connection to reclaim resources" - this might be something the `Service` knows about but the server transport code does not.
- "DNS servers can close the connection if the idle timeout set by local policy is exceeded" - timeout policy is handled by the server transport code, but presumably a service could itself decide a policy has been violated and wish to close a connection?
- "connections can be closed by either end under unusual conditions such as defending against an attack or system failure/reboot" - this is the malicious case I referred to above.

So, I am open to changing this, but would at least like to confirm given the above before I make changes.

Reply:
Point one is for the lower level transport code. That part knows about connections, Service doesn't. Point two is also for the low level code. Again, Service doesn't know about idle connections. Point 3 is unlikely to improve by closing a connection. The attacker can just open a new one. This also needs action at a low level. So I would conclude that at this state, a service is not going to ask for a connection to be closed.

Reply:
We can put this on the TODO list

Reply:
Imagine that the Service sits above multiple servers and sees abusive behaviour from a given IP on one of them, it might want to disconnect any connections from that IP on the other servers too, though that might be more a need to issue reconfiguration instructions to servers regarding their defensive posture? I'll add this all to the TODO list.

Reply:
Actually, for now I can just remove the connection close as you've repeatedly said you see no use for it, we can always add it back later if a need arises.

Reply:
I have removed it.